pkg/generic/proto/json.go

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2023 CloudWeGo Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package proto
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/cloudwego/kitex/pkg/remote/codec/perrors"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/dynamic"
+)
+
+// NewWriteJSON build WriteJSON according to ServiceDescriptor
+func NewWriteJSON(svc *desc.ServiceDescriptor, method string, isClient bool) (*WriteJSON, error) {
+	fnDsc := svc.FindMethodByName(method)
+	if fnDsc == nil {
+		return nil, fmt.Errorf("missing method: %s in service: %s", method, svc.GetFullyQualifiedName())
+	}
+
+	// create dynamic message, get InputType if client, get OutputType if server
+	msgDescriptor := fnDsc.GetInputType()
+	if !isClient {
+		msgDescriptor = fnDsc.GetOutputType()
+	}
+	inputMessage := dynamic.NewMessage(msgDescriptor)
+
+	ws := &WriteJSON{
+		dynamicMessage: *inputMessage,
+		isClient:       isClient,
+	}
+	return ws, nil
+}
+
+// WriteJSON implement of MessageWriter
+type WriteJSON struct {
+	dynamicMessage dynamic.Message
+	isClient       bool
+}
+
+var _ MessageWriter = (*WriteJSON)(nil)
+
+// Write writes to a dynamic message and returns an output bytebuffer
+func (m *WriteJSON) Write(ctx context.Context, msg interface{}) (interface{}, error) {
+	// msg is string
+	s, ok := msg.(string)
+	if !ok {
+		return nil, perrors.NewProtocolErrorWithType(perrors.InvalidData, "decode msg failed, is not string")
+	}
+
+	// Convert string into json
+	var jsonMap map[string]interface{}
+	err := json.Unmarshal([]byte(s), &jsonMap)
```

Comment:
The conversion from string to []byte results in a copy of all bytes, which may be expensive.

Reply:
You may use gjson (as in thrift/json.go:originalWrite), or try with the `unsafe string to byte slice` (google it)

Reply:
I used `unsafe string to byte slice` since since gjson doesn’t perform validation to ensure that the string is in the correct format!

Reply:
Actually, we have a method `StringToSliceByte` in kitex/pkg/utils/byte2str.go. You can use it.